doc/source/scipeps/1-Optimizer.rst

```diff
@@ -0,0 +1,981 @@
+``Optimizer.__next__``. At each iteration the solution state is available to
+the user, which can be used for many purposes including: user defined halting
+criteria, modification of solver hyper-parameters, tracking solution
+trajectories, etc. Running the optimizer to completion is achievable with the
+``solve`` or ``__call__`` methods.
+
+Different optimization algorithms can inherit from ``Optimizer``, with each of
+the subclasses overriding the ``__next__`` method to represent the core of their
+iterative technique. For some solvers, each iteration is implemented in
+C/Fortran with the main optimization loop in Python (e.g., LBFGSB). We are not
+proposing to replace those external calls at this time.
+
+Other optimizers run the complete optimization in external C/Fortran code
+(e.g., ``leastsq`` which calls ``minpack``). These methods can run the entire
+optimization in external code. Future work may involve performing a single
+optimization step in C/Fortran by extraction of iteration logic into Python or
+Cython, but is beyond the scope of this proposal (and would also require
+rigorous benchmarking and testing).
+
+The proposed changes will be transparent to an end-user of ``minimize`` and
+``fmin``. The intermediate or advanced user will appreciate the ``Optimizer``
+and ``Function`` classes.
+
+Timeline
+--------
+
+1. The Optimizer, Function classes are privately introduced, with subclasses
+   for some minimizers (NelderMead, LBFGS, BFGS, DifferentialEvolution) added.
+   They will be used as the core functionality of ``fmin``, etc.
+2. Subsequent private classes for the remaining minimizers will be created.
+   This will allow fine tuning of the Optimizer and Function classes.
+3. In later releases the ``Optimizer`` and ``Function`` classes will be
+   available in the SciPy public API.
+
+Rationale
+=========
+
+Simplified maintenance
+----------------------
+
+The maintenance burden of the new classes will be significantly reduced compared
+to the current state of scipy.optimize. It will be easier to develop new
+features and provide more comprehensive testing.
+
+The main reason for this is class inheritance. Improvements made to the base
+``Optimizer`` class mean that all that all inheriting objects improve. Currently
+such changes have to be made in each minimizer, which leads to code duplication,
+and the attendant risk of bugs being introduced.
+
+For example:
+
+* Placing numerical differentiation in the Function class
+    * would allow either absolute or relative delta change to be made easily,
+      and in one place.
+    * and, with the current API, require modifications and extra keywords for
+      all minimizer functions.
+* The user wishes to halt optimization early (issues `#4384
+  <https://github.com/scipy/scipy/issues/4384>`_, `#7306
+  <https://github.com/scipy/scipy/issues/7306>`_).
+    * This would be simply achieved in the new framework via
```

Comment:
some more housekeeping: this doesn't render quite right in https://github.com/andyfaff/scipy/blob/8fcd7d98c5fa1fddd765b08e4426d4b493f48420/doc/source/scipeps/1-Optimizer.rst

Reply:
Is there a quick way of rendering the document on my computer to see if I am improving things?

Reply:
There is if we'd actually add it to the html docs in this PR. Although rebuilding the docs isn't really faster than pushing to Github and viewing the reST rendering. There are also tools to locally view rendered reST, but I don't use those so don't have a good recommendation.

Reply:
You could paste your code into an online reST editor like http://rst.ninjs.org/ if you want to quickly check for formatting & rendering errors.

Reply:
Thanks for that @lagru. I think I resolved the issues in 00a6446852a0decb7f9c8e2de19f018b434ce2ad